plutus-core/plutus-core/src/PlutusCore/Parser/ParserCommon.hs

```diff
@@ -155,7 +167,17 @@ version = trailingWhitespace $ do
 
 -- | Parses a `Name`. Does not consume leading or trailing whitespaces.
 name :: Parser Name
-name = try $ do
-    void $ lookAhead letterChar
-    str <- takeWhileP (Just "identifier") isIdentifierChar
-    Name str <$> intern str
+name = try $ parseUnquoted <|> parseQuoted
+  where
+    parseUnquoted = do
+        void $ lookAhead (satisfy isIdentifierStartingChar)
+        str <- takeWhileP (Just "identifier-unquoted") isIdentifierChar
+        Name str <$> intern str
+    parseQuoted = do
+        void $ char backTick
+        void $ lookAhead (satisfy isQuotedIdentifierChar)
+        strUnquoted <- takeWhileP (Just "identifier-quoted") isQuotedIdentifierChar
+        void $ char backTick
+        let str = T.cons backTick (T.snoc strUnquoted backTick)
+        Name str <$> intern str
```

Comment:
I think I slightly disagree with this approach. This approach says that the parsed name is the string _including_ the backticks. That seems wrong to me: the backticks are part of  the _syntax_ of names, that allow us to give an _actual name_ which has special characters in it. The backticks themselves are not part of the name.

If we view them as syntax then we do things slightly differently:
- The name we parse here is without the backticks
- When we prettyprint a name, we need to decide what syntax to use depending on whether it can be printed unquoted or not

Reply:
Scala is a language I know that supports quoted names. I thought `foo` and `` `foo` `` are different names, but I was wrong - they are the same name. So I agree with you.

Reply:
@michaelpj Updated. The only slight problem is that `(all *_0 (type) *_0)` becomes `` (all `*`_0 (type) `*`_0) `` (see [e1b585e](https://github.com/input-output-hk/plutus/pull/5218/commits/e1b585eb9bbae8440b4e0321a72912d0a690cdf7)). The `*` comes from `dummyTyname`. Can we use something else there, say `a`? That would be less confusing anyway since `*` usually refers to kinds.

Reply:
> We can still use a for that. I think GHC invents type variables on similar situations, rather than using *

GHC makes _fresh_ type variables for each such thing, which is fine. We'd have to do something similar, which complicates things.

`any` seems okay.

> I think the behavior of Parser Name has always been wrong. If the name being parsed contains the unique. It would parse name_unique into the name field.

It's kind of dumb, but this is sort of fine when we use that syntax to print uniques: then it effectively produces globally unique names, which parses fine! Not sure if anyone relies on that, though :thinking:

Reply:
> GHC makes fresh type variables for each such thing, which is fine.

We can reserve a special unique for `dummyName`, but I'll just go with `any`.

> It's kind of dumb, but this is sort of fine when we use that syntax to print uniques: then it effectively produces globally unique names, which parses fine! Not sure if anyone relies on that, though thinking

:facepalm: 

Ok, then I think the easiest solution to restore old behavior is to quote around `name_unique`, i.e., `` `foo_i0` `` instead of `` `foo`_i0 ``

Reply:
I do agree it's super ugly, and this now just makes the ugliness very apparent... not that that helps :shrug:

Reply:
We have an [issue](https://github.com/input-output-hk/plutus/issues/4808) about pretty-printing of uniques. If anybody feels like taking over, that would be highly appreciated.

Reply:
Updated based on comments above.